azavea/echo-locator
# Enable user profile editing

## Overview

This PR allows users to edit their profiles from the frontend, rather than only in the admin. It adds a `saveProfile` action which sends a `PUT` request to the `UserProfileView` (previously named `GetUserProfile`) on save to update existing profiles and destinations and add new destinations. It also makes a few small tweaks, such as changing "Day care" (as it was saved in the backend) to "Daycare" (as it was used in the frontend) to fix potential key errors and adds the user's email (as read-only) to the `EditProfile` component.


### Notes

While I found that I could use the `UserProfileView` to update destinations and profiles and create new destinations, deleting destinations via that endpoint was harder and messier. So I've added a separate action and endpoint for deleting destinations that is triggered directly by the delete button next to each destination, as opposed to the profile/destination updates, which are sent to the backend all at once on profile save.

The hardest part of this card was state management, especially between the `EditProfile` component state, the app state, and the database. I've tried to minimize the places where these 3 can diverge from each other by only updating database state on profile save or destination delete. Initially, I had planned to let the existing `handleAuthChange` flow handle updating `userProfile` in app state; however, I changed it to update inside the `saveProfile` action after a successful database update instead in order to prevent app and database state from diverging. I also added an error message so that when updating the database fails, it at least notifies the user rather than just not saving their changes; I just couldn't figure out how to make that messaging genuinely helpful and say something other than "try again."


## Testing Instructions

 * Run `./scripts/update` to make the new "day care" to "daycare" migration, then start the server.
 * Login and select "Edit Profile". Verify that your profile looks right based on what's in the backend.
 * Delete a destination and verify that the destination no longer appears on your profile in the backend.
 * Edit one your existing destinations and add a new one; check that both appear as they should in the backend.
 * Make edits to the rest of your profile and verify that they show up as expected in the backend.


Resolves #443
